README.md
<img width="150" height="150" align="left" style="float: left; margin: 0 10px 0 0;" alt="Atlanta" src="https://i.goopics.net/lEENx.png">  

# Atlanta

[![](https://img.shields.io/discord/565048515357835264.svg?logo=discord&colorB=7289DA)](https://discord.atlanta-bot.fr)
[![](https://discordbots.org/api/widget/status/557445719892688897.svg)](https://discordbots.org/bot/557445719892688897)
[![](https://img.shields.io/badge/discord.js-v12.0.0--dev-blue.svg?logo=npm)](https://github.com/discordjs)
[![](https://img.shields.io/badge/patreon-donate-orange.svg)](https://www.patreon.com/androz2091)
[![](https://www.codefactor.io/repository/github/Androz2091/AtlantaBot/badge)](https://www.codefactor.io/repository/github/Androz2091/AtlantaBot)

> This bot is used by more than 90,000 Discord users and more than 500 servers.

Atlanta is a open source Discord bot coded in JavaScript with [Discord.js](https://discord.js.org) and [Mongoose](https://mongoosejs.com/docs/api.html) by [Androz2091](https://github.com/Androz2091).  
Feel free to add a star ⭐ to the repository to promote the project!

## Features

### Complete Bot

Atlanta offers (non-exhaustive list):
*   ✉️ Support for commands in direct messages
*   🇫🇷 Support for translations (illimited languages)
*   ⚙️ Guild configuration (prefix, ignored channels, etc...)
*   😀 Commands made pleasant thanks to the many emojis
*   🗳️ Support for Discordbots.org votes with rewards

Atlanta also adds **a new mention** like @everyone and @here, the `@someone`, which allows you to pick a random member of the server!

### Many commands

Atlanta has a lot of features, with **7 main categories**:

*   👩‍💼 **Administration**: `slowmode`, `welcome`, `addcommand`, `automod`, `backup` and **11** more! 
*   🚓 **Moderation**: `see-warns`, `setlogs`, `checkinvites`, `poll`, and **6** more! 
*   🎵 **Music**: `play`, `skip`, `queue`, `np`, and **2** more! 
*   💰 **Economy**: `profile`, `work`, `badge`, `slots`, `balance`, `leaderboard`, and **6** more! 
*   👻 **Fun**: `lmg`, `findwords`, `flip`, `lovecalc`, `random` and **6** more! 
*   🖨️ **General**: `minimize`, `setafk`, `translate`, `remindme`, `hastebin`, `fortnite`, `minecraft` and **10** more! 
*   👑 **Owner**: `eval`, `getinvite`, `servers-list` and **2** more!

[See the full commands list](https://www.atlanta-bot.fr/commands)

### A powerful Dashboard

Atlanta has its own dashboard which also offers many features! The dashboard runs with Express and EJS!

<img align="left" style="float: centrer; margin: 0 10px 0 0;" src="https://zupimages.net/up/19/31/c3ya.png" height="200" width="350"/>
<img align="center" style="float: left; margin: 0 10px 0 0;" src="https://zupimages.net/up/19/31/vnq5.png" height="200" width="350"/>
<img align="center" style="float: centrer; margin: 0 10px 0 0;" src="https://zupimages.net/up/19/31/htga.png" height="200" width="350"/>

You can directly **edit your configuration**, **manage your servers**, **view rankings**, **modify your profile** and much more!

> Find the code in the dashboard folder! 

## Installation

### Non-customized version

If you don't want to edit the code of the bot, a permanent online version is available, which you can invite to your own Discord!   

[![Discord Bots](https://discordbots.org/api/widget/557445719892688897.svg)](https://discordbots.org/bot/557445719892688897)

### Customized version

If you want to edit the bot's code and host it on your machine, it's possible!  
You must follow the [installation guide](https://www.atlanta-bot.fr/installation/) to properly install it! Don't worry, it's fast and simple if you follow the guide!

## Links

*   [Commands List](https://www.atlanta-bot.fr/commands)
*   [Installation Guide](https://www.atlanta-bot.fr/installation)
*   [F.A.Q](https://www.atlanta-bot.fr/faq/)
*   [Discord](https://discord.atlanta-bot.fr)
*   [Twitter](https://twitter.com/AtlantaBot)
*   [Github](https://github.com/Androz2091/AtlantaBot/)
*   [Dashboard](https://dashboard.atlanta-bot.fr)

## Contributing

Before **creating an issue**, please ensure that it hasn't already been reported/suggested, and double-check the [F.A.Q](https://www.atlanta-bot.fr/faq).   
And if you have a question, please ask it in the [Discord server](https://discord.atlanta-bot.fr/) instead of opening an issue.
If you wish to contribute to the Atlanta codebase or documentation, feel free to fork the repository and submit a pull request!

## License

Atlanta is licensed under the GPL 3.0 license. See the file `LICENSE` for more information. If you plan to use any part of this source code in your own bot, I would be grateful if you would include some form of credit somewhere.


docs/changelog.md
# Changelog

All notable changes to this project will be documented in this file.

## [4.7.0]

### Added

* Atlanta now uses i18next and can be translated on the [Crowdin project page](https://crowdin.com/project/atlantabot).
* Simple cache system was added, to reduce latency. (c73d1d6b)

### Fixed

* The play command is now playing the right song instead of a random one. (be4f497c)
* Dashboard authentication has been fixed. (ad3d84a6)
* Birthdate input on the dashboard has been fixed and is now saving the right date. (b885ccc7)
* `addemote` command has been renamed to `addemoji`. (e8b3b969)
* `tcl` command has been removed, as it was not used. (c3309663)
* `purge` command was removed, as it was not used. (07e0d55d)
* `api-token` command has been removed, as it was useless. (5857aab1)
* `partners` command has been removed, as it was not used. (dec093fa)
* `mention-role` command was removed, as Discord now has a mention permissions. (3fb32d7d)
* `getconf` and `getinvite` commands were temporarily removed. (aa998c58)
* `help` command now generate an invite link with administrator permissions. (b64114a6)
* `weeding` command has been renamed to `marry`. (dd89ab8b)
* Dashboard selector (the grid with servers) is now working for accounts with +30 servers. (2149b903)
* Dashboard now accepts prefixes of one character. (2fac3b82)
* New `h` and `commands` aliases to `help` command. (081d0ead)
* New `lc` alias to `lovecalc` command. (8e612a4a)
* `userinfo` command now sends an error message if the provided ID is not valid. (9ebfd866)
* `skip` command is now working when the voice channel includes bots that can't vote. (095dac3d)
* Atlanta now uses the `discord-canvas` library for fortnite related commands and greeting images. (aaf785b5)
* Atlanta now uses the `moment` library to show dates and times. (f46edf00)
* Atlanta now uses the last version of the `discord.js` library. (b33f4e20)
* Atlanta now uses the last version of the `discord-giveaways` library. (1a57cebb)
* Atlanta now uses the last version of the `discord-player` library (YouTube API key is no longer required). (80335aa8)
* Sentry.io has been setup, and errors are now saved and reported. (8bfbb03a)

## [4.6.0]

### Added

* `serverinfo` command now displays the number of boosts of the server
* New `achievements` command
* `images` commands have been fixed

## [4.5.10]

### Added

* New `mention-role` command

### Fixed

* If a member is muted and leaves and returns, he will always be muted
* User game is now displayed in the userinfo command
* Fix `invitations` command
* Rep cooldown is now user based and not member based
* `steal` command was renamed to `rob`
* `minimize` command was renamed to `shorturl`
* `ban` command now accepts ids
* Fixes some other minor bugs

## [4.5.0]

### Added

* New `fortniteshop` command
* New Fornite Shop channel, in which a shop image made with canvas is sent every morning
* New `staff` command to get the staff members list of a guild
* New `deletemod` command if you want that the moderation commands will be deleted automatically
* New `clearwarns` command to delete all warns of a member

### Dashboard

* New `Statistics` tab
    * New `Guild Stats` page to get the stats of a specific guild
    * New `Stats` button on the selector page
* New `settings` page to configure your profile and the website (language, logout, etc...)

### Economy

* Economy is now guild based
* New `steal` command
* New `deposit` command
* New `withdraw` command
* `profile` command was edited
* `work` and `rep` cooldown is now 6h instead of 12h
* Add `work` streak

## [4.0.0] - 2019-01-08

### Added

* Welcome and goodbye images were improved (using Canvas)
* Fortnite stats are now displayed in images (using Canvas)
* New `calc` command
* 13 new commands in `images` category
* `*clear all` to clear all the messages in a channel
* `vote` alias has come back

### Fixed

* Now no longer considers that you are over 80 years old with the `birthdate` command
* `clear` command now checks if the parameter is a number
* `announcement` command now checks if the message length is higher than 1030
* `warn` command now checks if the mentionned member is not a bot
* `setlogs` command renamed to `setmodlogs`
* `ignore`, `slowmode`, `automod`, `goodbye`, `welcome`, `setmodlogs` and `setsuggests` now check if the mentionned channel is a text channel and if the channel is in the current guild
* The design of the `help` command has been improved
* The `help` command now displays the number of commands executed in the last 7 days
* And many other small bugs!

### Dashboard

* Dashboard is back in the same repository as Atlanta to make things easier
* Support for almost all server settings (welcome messages, goodbye messages, prefix, language, etc...)
* English **and** French support
* Page for displaying the ranking
* Page to display a user's profile
* Page to edit the profile
* Server search form

## [3.5.0] - 2019-11-07

### Added

* New `backup` command
* New `announcement` command
* New `giveaway` command
* Users commands cooldown
* Vote system for `skip` and `stop` commands
* `-g` option to the help command
* Translation for commands usage and commands examples
* The `minecraft` command supports Minecraft Pocket Edition servers

### Changed

* Now use Discord.js v12.0.0 instead of Discord.js v11.5.0
* Now use DeepL Translate instead of Yandex Translate
* Now use MongoDB instead of quick.db
* Now use Node-Fetch instead of Snekfetch
* Now use of `let` instead of `var` for a better scoping
* Clarification of language files
* Discordbots.org vote is now rewarded with 40 credits
* Most commands are now available in private messages
* The `help` command is now paginated
* `leave` command renamed to `goodbye` command
* `credits` command renamed to `money` command
* `warns` command renamed to `sanctions` command
* All the emojis were updated
* Update links in the `invite` command
* Best parse of time

### Removed


base/User.js
const mongoose = require("mongoose"),
	Canvas = require("canvas");

const genToken = () => {
	let token = "";
	const characters = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwzy0123456789.-_";
	for (let i = 0; i < 32; i++){
		token += characters.charAt(Math.floor(Math.random() * characters.length));
	}
	return token;
};

const userSchema = new mongoose.Schema({

	/* REQUIRED */
	id: { type: String }, // Discord ID of the user

	/* ECONOMY (GLOBAL) */
	rep: { type: Number, default: 0 }, // Reputation points of the user
	bio: { type: String }, // Biography of the user
	birthdate: { type: Number }, // Birthdate of the user (the timestamp)
	lover: { type: String }, // The person with whom the user is in a relationship

	/* STATS */
	registeredAt: { type: Number, default: Date.now() }, // Registered date of the user

	/* ACHIEVEMENTS */
	achievements: { type: Object, default: {
		married: {
			achieved: false,
			progress: {
				now: 0,
				total: 1
			}
		},
		work: {
			achieved: false,
			progress: {
				now: 0,
				total: 10
			}
		},
		firstCommand: {
			achieved: false,
			progress: {
				now: 0,
				total: 1
			}
		},
		slots: {
			achieved: false,
			progress: {
				now: 0,
				total: 3
			}
		},
		tip: {
			achieved: false,
			progress: {
				now: 0,
				total: 1
			}
		},
		rep: {
			achieved: false,
			progress: {
				now: 0,
				total: 20
			},
		},
		invite: {
			achieved: false,
			progress: {
				now: 0,
				total: 1
			}
		}
	}},

	/* COOLDOWN */
	cooldowns: { type: Object, default: {
		rep: 0
	}},

	/* OTHER INFORMATIONS */
	afk: { type: String, default: null }, // Whether the member is AFK
	reminds: { type: Array, default: [] }, // the reminds of the user
	logged: { type: Boolean, default: false }, // if the user is logged to the dashboard
	apiToken: { type: String, default: genToken() } // the api token of the user

});

userSchema.method("genApiToken", async function(){
	this.apiToken = genToken();
	await this.save();
	return this.apiToken;
});

userSchema.method("getAchievements", async function(){
	const canvas = Canvas.createCanvas(1800, 250),
		ctx = canvas.getContext("2d");
	const images = [
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.work.achieved ? "_colored" : "")+"1.png"),
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.firstCommand.achieved ? "_colored" : "")+"2.png"),
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.married.achieved ? "_colored" : "")+"3.png"),
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.slots.achieved ? "_colored" : "")+"4.png"),
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.tip.achieved ? "_colored" : "")+"5.png"),
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.rep.achieved ? "_colored" : "")+"6.png"),
		await Canvas.loadImage("./assets/img/achievements/achievement"+(this.achievements.invite.achieved ? "_colored" : "")+"7.png")
	];
	let dim = 0;
	for(let i = 0; i < images.length; i++){
		await ctx.drawImage(images[i], dim, 10, 350, 200);
		dim += 200;
	}
	return canvas.toBuffer();
});

module.exports = mongoose.model("User", userSchema);